main-board/src/main.rs

```diff
@@ -108,99 +58,53 @@ fn main() -> ! {
+
+    /* Setup peripherals */
+    let pads = hal::sercom::uart::Pads::<hal::sercom::Sercom5, _>::default()
+        .rx(pins.pb17)
+        .tx(pins.pb16);
+    let mut uart = hal::sercom::uart::Config::new(
         &peripherals.MCLK,
-        peripherals.SERCOM1,
-        pads_spi,
-        spi_clk.freq(),
+        peripherals.SERCOM5,
+        pads,
+        uart_clk.freq(),
+    )
+    .baud(
+        9600.hz(),
+        hal::sercom::uart::BaudMode::Fractional(hal::sercom::uart::Oversampling::Bits16),
     )
-    .length::<hal::sercom::spi::lengths::U1>()
-    .bit_order(hal::sercom::spi::BitOrder::MsbFirst)
-    .spi_mode(hal::sercom::spi::MODE_0)
     .enable();
-    /* End SPI config */
-
-    /* Start sd controller config */
-    let time_sink: TimeSink = TimeSink::new();
-    let mut sd_cont = sd::Controller::new(sd::SdMmcSpi::new(sdmmc_spi, cs), time_sink);
-    match sd_cont.device().init() {
-        Ok(_) => match sd_cont.device().card_size_bytes() {
-            Ok(_) => nb::block!(uart.write(b'1')).unwrap(),
-            Err(_) => nb::block!(uart.write(b'0')).unwrap(),
-        },
-        Err(_) => {
-            for b in b"Error\r\n" {
-                nb::block!(uart.write(*b)).unwrap();
-            }
-        }
-    }
-
-    let mut volume = match sd_cont.get_volume(sd::VolumeIdx(0)) {
-        Ok(volume) => volume,
-        Err(_) => {
-            for byte in b"Error getting volume!\r\n" {
-                nb::block!(uart.write(*byte)).unwrap();
-            }
-            panic!("Cannot get volume 0");
-        }
-    };
-
-    let root_directory = match sd_cont.open_root_dir(&volume) {
-        Ok(root_directory) => root_directory,
-        Err(_) => {
-            for byte in b"Error getting root!\r\n" {
-                nb::block!(uart.write(*byte)).unwrap();
-            }
-            panic!("Cannot get root");
-        }
-    };
-
-    let file = sd_cont.open_file_in_dir(
-        &mut volume,
-        &root_directory,
-        "test.txt",
-        sd::Mode::ReadWriteCreateOrTruncate,
+    let mut micro_sd = sd_interface::SdInterface::new(
+        &peripherals.MCLK, 
+        peripherals.SERCOM1, 
+        spi_clk, 
+        pins.pa18.into_push_pull_output(),
+        pins.pa17.into_push_pull_output(),
+        pins.pa19.into_push_pull_output(),
+        pins.pa16.into_push_pull_output()
     );
-    let mut file = match file {
-        Ok(file) => file,
-        Err(_) => {
-            for byte in b"Error creating file!\r\n" {
-                nb::block!(uart.write(*byte)).unwrap();
-            }
-            panic!("Cannot create file.");
-        }
+    /* Test peripherals */
+    let mut test_file = micro_sd.open_file("testing.txt");
+    micro_sd.write(&mut test_file, b"Hello this is a test!");
+    micro_sd.write_str(&mut test_file, "Testing Strings");
+    micro_sd.close_file(test_file); // we are done with the file so destroy it
+    micro_sd.close(); // we are done our test so destroy
```

Comment:
Hum, it would be nice to get those tests working (similar to https://ferrous-systems.com/blog/test-embedded-app/). I wonder if we should start in this PR.

Reply:
I would like to start proper testing in another PR.